### PR TITLE
Add `--skip-if-loaded` for `ecto.load`

### PR DESCRIPTION
**Pull request by ulissesalmeida**
Today, if we want to add the `mix ecto.load`
to a pipeline of tasks, it will fail if the
database structure is already up and it will
not run the next tasks.

For example, let's say I want to build a
pipeline called `ecto.setup` like this:

```
defp aliases do
    [
      "ecto.setup": [
        "ecto.create",
        "ecto.load",
        "ecto.migrate",
        "run priv/repo/seeds.exs"
    ],
```
1. If the database is already up, the `ecto.create`
doesn't fail and proceed to `ecto.load`. 
2. In `ecto.load`, fails when there are tables there and doesn't
execute the next tasks in the pipeline. 

Unfortunately, having tables in the database doesn't mean problem in this
pipeline. It would be cool just skip the `ecto.load` if the tables are already there.

We discussed this option here in the mailing list:
https://groups.google.com/forum/#!topic/elixir-ecto/8aw2DL_c9Yc

And also in the PR here:
https://github.com/elixir-ecto/ecto_sql/pull/108

That introduced the new feature here:
https://github.com/elixir-ecto/ecto_sql/pull/111

We discussed that a good approach would be, before attempting to
load the dump, would be check if there's the migration table there. If yes,
it's not a new database, it's loaded. If there are no migrations table there, , we can
try to load the dump.

With the `--skip-if-loaded`, when the database is not new, we
can skip the load and succeed in the task.

### CODE REVIEW

File: test/mix/tasks/ecto.dump_load_test.exs [hunk 1]
```diff
@@ -41,7 +45,7 @@ defmodule Mix.Tasks.Ecto.DumpLoadTest do
   end
 
   setup do
-    Application.put_env(:ecto_sql, __MODULE__.Repo, [])
+    Application.put_env(:ecto_sql, __MODULE__.Repo, priv: "tmp/#{inspect(Ecto.Load)}")
```
Comment (ulissesalmeida) [inline test/mix/tasks/ecto.dump_load_test.exs]: oops, we don't need this anymore 😊

File: lib/mix/tasks/ecto.load.ex [hunk 2]
```diff
@@ -51,29 +52,46 @@ defmodule Mix.Tasks.Ecto.Load do
...
-                                          "load structure for #{inspect repo}")
+      ensure_implements(
+        repo.__adapter__,
+        Ecto.Adapter.Structure,
+        "load structure for #{inspect(repo)}"
+      )
+
+      {:ok, pid, _} = ensure_started(repo, all: true)
+
+      loaded? = table_exists?.(repo, "schema_migrations")
```
[the remark this answers is not in the document]
Comment (josevalim) [reply lib/mix/tasks/ecto.load.ex]:
```suggestion
      loaded? = table_exists?.(repo, Ecto.Migration.SchemaMigration.get_source(repo))
```

File: test/mix/tasks/ecto.dump_load_test.exs [hunk 3]
```diff
@@ -10,14 +10,18 @@ defmodule Mix.Tasks.Ecto.DumpLoadTest do
...
 
     defmacro __before_compile__(_), do: :ok
-    def dumpers(_, _), do: raise "not implemented"
-    def loaders(_, _), do: raise "not implemented"
-    def init(_), do: raise "not implemented"
-    def checkout(_, _, _), do: raise "not implemented"
-    def ensure_all_started(_, _), do: raise "not implemented"
+    def dumpers(_, _), do: raise("not implemented")
+    def loaders(_, _), do: raise("not implemented")
+    def checkout(_, _, _), do: raise("not implemented")
```
Comment (josevalim) [inline test/mix/tasks/ecto.dump_load_test.exs]: Those changes are unrelated, please revert. :)

File: lib/mix/tasks/ecto.load.ex [hunk 4]
```diff
@@ -51,29 +52,46 @@ defmodule Mix.Tasks.Ecto.Load do
...
 
-      if skip_safety_warnings?() or
-          opts[:force] or
-          Mix.shell.yes?("Are you sure you want to load a new structure for #{inspect repo}? Any existing data in this repo may be lost.") do
-        load_structure(repo, opts)
+      cond do
+        loaded? and opts[:skip_if_loaded] ->
+          Mix.shell().info("Skipped since a structure for #{inspect(repo)} has been loaded")
+
+        (skip_safety_warnings?() and !loaded?) or opts[:force] or confirm_load(repo) ->
```
[the remark this answers is not in the document]
Comment (josevalim) [reply lib/mix/tasks/ecto.load.ex]: Now that I think about it, here is what need:

1. If the repository is loaded, we will ask if they want to proceed
2. If the safety warnings are enabled, we will ask if they want to proceed

We can use :force to override both. So we would do this:

```
if (skip_safety_warnings?() and not loaded?) or opts[:force] or confirm_load(repo, loaded?) do
  load_structure(repo, opts)
```
  
And we change the message we ask users on `confirm_load` depending if `loaded?` is true or false. Wdyt?

**Comment by josevalim** [comment]
@ulissesalmeida I have added some comments. We are almost there. Can you also please rebase? Thank you! :heart:

File: lib/mix/tasks/ecto.load.ex [hunk 5]
```diff
@@ -51,29 +52,46 @@ defmodule Mix.Tasks.Ecto.Load do
...
+
+      pid && repo.stop()
 
-      if skip_safety_warnings?() or
-          opts[:force] or
-          Mix.shell.yes?("Are you sure you want to load a new structure for #{inspect repo}? Any existing data in this repo may be lost.") do
-        load_structure(repo, opts)
+      cond do
+        loaded? and opts[:skip_if_loaded] ->
+          Mix.shell().info("Skipped since a structure for #{inspect(repo)} has been loaded")
```
Comment (josevalim) [inline lib/mix/tasks/ecto.load.ex]:
```suggestion
          Mix.shell().info("Skipping ecto.load since a structure for #{inspect(repo)} has been loaded")
```

Comment (josevalim) [inline lib/mix/tasks/ecto.load.ex]: Since you are explicitly giving skip_if_loaded, I think we can just return `:ok`?

```suggestion
          :ok
```

File: lib/mix/tasks/ecto.load.ex [hunk 6]
```diff
@@ -51,29 +52,46 @@ defmodule Mix.Tasks.Ecto.Load do
...
+
+        (skip_safety_warnings?() and !loaded?) or opts[:force] or confirm_load(repo) ->
+          load_structure(repo, opts)
       end
```
Comment (josevalim) [inline lib/mix/tasks/ecto.load.ex]: I think we need a pokemon clause here that does:

```
true ->
  :ok
```

File: lib/mix/tasks/ecto.load.ex [hunk 7]
```diff
@@ -51,29 +52,56 @@ defmodule Mix.Tasks.Ecto.Load do
...
+  defp confirm_load(repo, false) do
+    Mix.shell().yes?(
+      "Are you sure you want to load a new structure for #{inspect(repo)}? Any existing data in this repo may be lost."
+    )
+  end
+
+  defp confirm_load(repo, true) do
+    Mix.shell().yes?("""
+    It looks like a structure was already loaded for #{inspect(repo)}. Any attempt to load it again might fail.
+    Are you sure you want to proceed?
```
Comment (ulissesalmeida) [inline lib/mix/tasks/ecto.load.ex]: @josevalim what do you think about this message?

File: lib/mix/tasks/ecto.load.ex (hunk 4 above)
[the remark this answers is not in the document]
Comment (ulissesalmeida) [reply lib/mix/tasks/ecto.load.ex]: Sounds good! ✨

File: lib/mix/tasks/ecto.load.ex [hunk 8]
```diff
@@ -51,29 +52,56 @@ defmodule Mix.Tasks.Ecto.Load do
...
       ensure_repo(repo, args)
-      ensure_implements(repo.__adapter__, Ecto.Adapter.Structure,
-                                          "load structure for #{inspect repo}")
+      ensure_implements(
+        repo.__adapter__,
+        Ecto.Adapter.Structure,
+        "load structure for #{inspect(repo)}"
+      )
+
+      {:ok, pid, _} = ensure_started(repo, all: true)
```
Comment (ulissesalmeida) [inline lib/mix/tasks/ecto.load.ex]: 🚨Now the `ensure_started` doesn't exist anymore. How can I ensure that 
repo is up before query `table_exists?` 

I could see that `ensure_started ` was moved to the `Migrator.with_repo/3`, but I don't think calling `Migrator` here would be cool.

Comment (josevalim) [reply lib/mix/tasks/ecto.load.ex]: It is totally fine to call the migrator here. :)

Comment (josevalim) [reply lib/mix/tasks/ecto.load.ex]: Especially because the function we are calling to check for tables is a migration table. :)

File: lib/mix/tasks/ecto.load.ex (hunk 7 above)
Comment (josevalim) [reply lib/mix/tasks/ecto.load.ex]: Beautiful.

File: lib/mix/tasks/ecto.load.ex (hunk 8 above)
Comment (ulissesalmeida) [reply lib/mix/tasks/ecto.load.ex]: Done! ✨

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
